controller/backing_image_controller.go

```diff
@@ -0,0 +1,621 @@
+		if diskStatus.DiskUUID == diskUUID {
+			if diskSpec, exists := node.Spec.Disks[id]; exists {
+				diskPath = diskSpec.Path
+			}
+			break
+		}
+
+	}
+	if diskPath == "" {
+		return fmt.Errorf("cannot find the disk path with UUID %v on node %v", diskUUID, node.Name)
+	}
+
+	// The image is not important.
+	// As long as the image is a Linux distribution with wget installed, it can be used in the download pod.
+	// We choose the default engine image here so that users don't need to fetch on extra image in the air-gaped environment.
+	defaultEngineImage, err := bic.ds.GetSetting(types.SettingNameDefaultEngineImage)
+	if err != nil {
+		return err
+	}
+	if defaultEngineImage.Value == "" {
+		return fmt.Errorf("default engine image not set")
+	}
+
+	filePathInContainer := path.Join(types.BackingImageDirectoryInContainer, types.BackingImageFileName)
+	downloadedNotificationFlag := path.Join(types.BackingImageDirectoryInContainer, "downloaded")
+	cmdArgs := []string{
+		"/bin/bash", "-c",
+		"rm " + filePathInContainer + " " + downloadedNotificationFlag + "; " +
+			"wget --no-check-certificate -c " + backingImage.Spec.ImageURL + " -O " + filePathInContainer + " > /dev/null 2>&1; " +
+			"if [ $? -eq 0 ]; then touch " + downloadedNotificationFlag + " && sync && echo downloaded; fi && " +
+			"trap 'rm " + filePathInContainer + " " + downloadedNotificationFlag + " && echo cleaned up' EXIT && sleep infinity"}
+
+	privileged := true
+	hostToContainerPropagation := corev1.MountPropagationHostToContainer
+	podManifest := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            podName,
+			Namespace:       bic.namespace,
+			OwnerReferences: datastore.GetOwnerReferencesForBackingImage(backingImage),
+			Labels:          types.GetBackingImageLabels(backingImage.Name, diskUUID),
+			Annotations:     map[string]string{types.GetLonghornLabelKey(types.LastAppliedTolerationAnnotationKeySuffix): string(tolerationsByte)},
+		},
+		Spec: corev1.PodSpec{
+			ServiceAccountName: bic.serviceAccount,
+			Tolerations:        util.GetDistinctTolerations(tolerations),
+			PriorityClassName:  priorityClass.Value,
+			Containers: []corev1.Container{
+				{
+					Name:            podName,
+					Image:           defaultEngineImage.Value,
+					ImagePullPolicy: imagePullPolicy,
+					Command:         cmdArgs,
+					SecurityContext: &corev1.SecurityContext{
+						Privileged: &privileged,
+					},
+					VolumeMounts: []corev1.VolumeMount{
+						{
+							MountPath:        types.BackingImageDirectoryInContainer,
+							Name:             "backing-image",
+							MountPropagation: &hostToContainerPropagation,
+						},
+					},
+					StartupProbe: &corev1.Probe{
+						Handler: corev1.Handler{
+							Exec: &corev1.ExecAction{
+								Command: []string{
+									"bash", "-c",
+									"ls " + filePathInContainer + " && ls " + downloadedNotificationFlag,
+								},
+							},
+						},
+						// Wait up to 1 day for the backing image downloading.
```

Comment:
1 day seems like a really long time to wait for the download?
Is it better to make this a shorter time, so we eventually mark the pod as failed which will mark the download as failed + cleanup so we retry?

Reply:
Sure, I have updated it to 3 hours